pallets/xyk/src/lib.rs

```diff
@@ -2167,41 +2225,87 @@ impl<T: Config> XykFunctionsTrait<T::AccountId> for Pallet<T> {
 	}
 
 	fn claim_rewards(
-		sender: T::AccountId,
-		liquidity_token_id: Self::CurrencyId,
-		amount: Self::Balance,
+		user: T::AccountId,
+		liquidity_asset_id: Self::CurrencyId,
+		mangata_amount: Self::Balance,
 	) -> DispatchResult {
 		let mangata_id: TokenId = T::NativeCurrencyId::get();
 
-		let current_block_number =
-			<frame_system::Pallet<T>>::block_number().saturated_into::<u32>();
-		let (rewards_total_user, rewards_claimed_user) = Pallet::<T>::calculate_rewards_amount(
-			sender.clone(),
-			liquidity_token_id,
-			current_block_number,
+		let (current_rewards, burned_not_claimed_rewards) =
+			Pallet::<T>::calculate_rewards_amount(user.clone(), liquidity_asset_id)?;
+
+		let total_claimable_rewards = current_rewards + burned_not_claimed_rewards;
+		ensure!(mangata_amount <= total_claimable_rewards, Error::<T>::NotEnoughtRewardsEarned);
+
+		// user is taking out rewards from LP which was already removed from pool
+		if mangata_amount <= burned_not_claimed_rewards {
+			let burned_not_claimed_rewards_new = burned_not_claimed_rewards - mangata_amount;
+			LiquidityMiningUserToBeClaimed::<T>::insert(
+				(&user, liquidity_asset_id),
+				burned_not_claimed_rewards_new,
+			);
+		}
+		// user is taking out more rewards then rewards from LP which was already removed from pool, additional work needs to be removed from pool and user
+		else {
+			LiquidityMiningUserToBeClaimed::<T>::insert((&user, liquidity_asset_id), 0 as u128);
+			// rewards to burn on top of rewards from LP which was already removed from pool
+			let rewards_to_burn = mangata_amount - burned_not_claimed_rewards;
+
+			let (
+				current_time,
+				user_work_total,
+				user_missing_at_checkpoint,
+				pool_work_total,
+				pool_missing_at_checkpoint,
+			) = Self::calculate_liquidity_checkpoint(user.clone(), liquidity_asset_id, 0 as u128)?;
+
+			let work_to_burn = multiply_by_rational(
+				Balance::try_from(user_work_total)?,
+				rewards_to_burn,
+				current_rewards,
+			)
+			.map_err(|_| Error::<T>::UnexpectedFailure)?;
+
+			LiquidityMiningUser::<T>::insert(
+				(&user, &liquidity_asset_id),
+				(current_time, user_work_total - work_to_burn, user_missing_at_checkpoint),
```

Comment:
please use checked operator, so in future noone will have to think if its gonna explode ;)

Reply:
let new_work_user = user_work_total.checked_sub(work_to_burn).ok_or_else(|| DispatchError::from(Error::<T>::NotEnoughtRewardsEarned))?;
			let new_work_pool = pool_work_total.checked_sub(work_to_burn).ok_or_else(|| DispatchError::from(Error::<T>::NotEnoughtRewardsEarned))?;
			
			LiquidityMiningUser::<T>::insert(
				(&user, &liquidity_asset_id),
				(current_time, new_work_user, user_missing_at_checkpoint),
			);
			LiquidityMiningPool::<T>::insert(
				&liquidity_asset_id,
				(current_time, new_work_pool, pool_missing_at_checkpoint),
			);